### ISSUE TITLE
Add multiple outlines, multiple drop shadows and gradients to DynamicFont

### ISSUE THREAD

**Issue by jesperkondrup** [godotengine/godot-proposals#2564]
### Describe the project you are working on
We are working on a game that uses text in many places with various effects applied like outlines and gradient. 
These texts should be able to change run-time, so it has to be a font instead of an image (see Photoshop image below).

![image](https://user-images.githubusercontent.com/24936603/113722450-b9e45b00-96f0-11eb-864f-74000e0e8f5f.png)

### Describe the problem or limitation you are having in your project
Currently the DynamicFont only supports 1 outline, no gradient and only 1 dropshadow. This means that we cannot create the same effect as in Photoshop where we often use 3 outlines, a gradient and some times a drop shadow. (The image below is what Godot is currently capable of).

![image](https://user-images.githubusercontent.com/24936603/113832378-dcc24e00-9788-11eb-88bd-f998f2319c73.png)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
 
 The solution could be to add the following options for the DynamicFont: 
 - Extra outlines (at least 3).
 - Gradient texture as color.
 - A second drop shadow.
 - [Bonus feature if possible] Soft drop shadow

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

In this mockup solution we have added 3 outlines, a gradient texture and an extra drop shadow.

![image](https://user-images.githubusercontent.com/24936603/113866828-0fcb0880-97ae-11eb-8f5f-b0a25ef3f759.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

We tried a solution our selves using a custom shader we hacked together by looking at older github posts but it did not work as intended. We got the gradient working though but not together with the outline and we still need to be able to create the other outlines as well. So it does not seem like this is possible.

![image](https://user-images.githubusercontent.com/24936603/113868161-aa781700-97af-11eb-89a2-7ffcfd881431.png)

### Is there a reason why this should be core and not an add-on in the asset library?

It would make sense to have this as a core feature because it's already there, it just need a bit more work before you can make some really cool text effects. This would also make creating fonts a lot easier for beginners as they can just play with the DynamicFont instead of spending many hours in Photoshop and Shoebox to make a simple Bitmap font. 

Others have requested a similar solution;
https://github.com/godotengine/godot/issues/19356

**Comment by Calinou** [comment]
See also https://github.com/godotengine/godot-proposals/issues/1963.

**Comment by jesperkondrup** [comment]
> Did you try increasing **Extra Spacing > Char** in the DynamicFont?

Ahh yes, I have removed it from my post. When I started writing I was not aware of extra spacing but discovered it later and yes the extra spacing works perfectly.

**Comment by 2plus2makes5** [comment]
Yes this would be awesome.

**Comment by Calinou** [comment]
@2plus2makes5 Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by 2plus2makes5** [comment]
Sorry for the last comment.

This issue made me realize how much can be done with texts so i decided to create workarounds to get what you wanted and i think i got really close(aside from the font and exact sizes and colors the only real difference is the inner outline of the first label being influenced by the gradient of the font):
![elaborate_texts](https://user-images.githubusercontent.com/35235318/114162425-5f820f00-9929-11eb-880a-38524b5c9a22.png)

It's definitively overkill but it's a simple solution that you can use as a temporary one at least, basically instead of using a single label with multiple outlines or shadows you use "layers" of labels, one over the other, each one with a different outline or shadow.

I also made a super simple shader for the gradient, i don't even know why the uv.y must be multiplied for 10 but it works so who cares lol, if you use it you must leave the font color white though.

**Comment by Zylann** [comment]
I'd like to point out that in your initial example, you describe a gradient being applied on the font resource, but on two screenshots, it is interchangeably shown spreading over the entire text, or just one line. How this would work is not explained.

With #51295 it would be possible to achieve the effect on a single line, but to make it spread on the whole text you either would need two fonts or a custom shader.

**Comment by jesperkondrup** [comment]
@Zylann Yes good point. Initially I think it would be fine if it just worked for 1 line. We also tried @2plus2makes5 approach which did seem to work but it was far too costly compared to just a sprite unfortunately.

**Comment by jdozubko** [comment]
This thread hasn't been mentioned in 2 years. Just wanted to bump it back up to relevance. Having font gradients would be amazing. I've been messing around with shaders trying to get a fix but the inconsistent UV coordinate mappings for font letters make it a nightmare. Building this feature in I think would be very welcome.

**Comment by KStormStudio** [comment]
So where can we just download this shader please?

**Comment by snowflower** [comment]
Is there a reason why we couldnt have another UV set of data (UV2) for each font glyph rendered?
So the UV2.y coordinate would be aligned to the baseline of the font text.
Above the baseline the V values would be positive, with the parts of the glyph under the baseline having a negative V value
It would make vertical font gradient mapping very easy.

**Comment by Delsin-Yu** [comment]
I'm thinking about implementing this feature to a dedicated variation of `ArtLabel`

**Comment by Calinou** [comment]
> I'm thinking about implementing this feature to a dedicated variation of `ArtLabel`

I'm not sure if this should be a dedicated node, as this would require duplicating lots of Label code. Also, none of its functionality would be accessible in RichTextLabel, let alone Label3D.

**Comment by Delsin-Yu** [comment]
Too bad that theme does not have packed array support

**Comment by Delsin-Yu** [comment]
`Label` can achieve the OP's requirement in the linked PR. I think I will implement the Gradient support for RichTextLabels.

![Image](https://github.com/user-attachments/assets/b6f1d295-9afa-4bed-bf8f-cf10d4fff9dc)

**Comment by jesperkondrup** [comment]
@Delsin-Yu Looks great. How did you achieve those extra outlines with the label node, is that using a theme?

**Comment by Delsin-Yu** [comment]
> is that using a theme?

Yes

**Comment by sunDalik** [comment]
@Delsin-Yu Can I ask how exactly did you achieve the third effect (where you have an outline over a shadow) with `LabelSettings`? As far as I can tell there's no way to change how outline is ordered relative to the shadow and they are always rendered independently.

**Comment by Delsin-Yu** [comment]
@sunDalik 

The trick is you don't use an outline for this case; you stack shadows.

<img width="2313" height="914" alt="Image" src="https://github.com/user-attachments/assets/2251e327-90ea-4718-b4e2-6817055bc3f9" />

One of the questions I brought up while developing this feature is the fact that stack shadows cover all the features stacked outlines have, but I ended up preserving both, as stacked outlines have better UX in their suited scenarios.

<img width="1459" height="408" alt="Image" src="https://github.com/user-attachments/assets/f18b8de0-2179-4f8a-a6e8-ab0f79354dba" />

**Comment by sunDalik** [comment]
@Delsin-Yu  Thank you, stacking shadows looks good although imperfect.  
I do agree that shadows can replace outlines but as of Godot 4.5 shadows don't stack together the same way outlines stack. Like, you can create 2 stacked outlines of size 8 and the total outline size will be 16. But if you create 2 stacked shadows, they don't actually stack and just render independently without changing based on previous shadows' sizes. Therefore, in order to achieve the effect in the image below (from godot 3) you need to precisely change shadow offsets and outline sizes and it's pretty finicky and inaccurate.

<img width="489" height="477" alt="Image" src="https://github.com/user-attachments/assets/3797c124-76eb-4ed2-822a-81209c3ba4d1" />

**Comment by Delsin-Yu** [comment]
@sunDalik I believe all you need is to add together the outline size of the previous one and the size you wish for the current one (e.g., if the previous one has an outline size of 5, and you want the current one to have an outline size of 3, you do `5+3=8`), this is also the calculation stacked outline uses.

**Comment by sunDalik** [comment]
@Delsin-Yu 
Here's what I mean: if you change the shadow offset, that changes the text's "shape" and the next shadow should adapt to it. Like in the image below I no longer have just a "single ring shape" I have "two rings" and the second shadow should also be in the shape of two rings. (Second image is how I would want it to work)
<img width="755" height="468" alt="Image" src="https://github.com/user-attachments/assets/fc036f78-a9eb-4889-aa45-2bbc52283e7b" /> 

<img width="199" height="191" alt="Image" src="https://github.com/user-attachments/assets/13e3c8fa-bf23-42db-b6b6-30032ee3c724" />

**Comment by Delsin-Yu** [comment]
@sunDalik Then all you need is to simply calculate them separately; stacked shadows are drawn in the order you defined them.

<img width="1251" height="815" alt="Image" src="https://github.com/user-attachments/assets/81c2de22-53dd-4cd0-ab59-f57bda14b8bc" />

**Comment by sunDalik** [comment]
@Delsin-Yu Oh my god thank you so much, that's exactly what I needed!
